src/hotspot/share/cds/heapShared.cpp

```diff
@@ -420,8 +418,13 @@ void HeapShared::archive_java_mirrors() {
       }
     }
   }
+}
 
-  delete_seen_objects_table();
+void HeapShared::archive_strings() {
+  oop shared_strings_array = StringTable::init_shared_table(_dumped_interned_strings);
+  bool success = archive_reachable_objects_from(1, _default_subgraph_info, shared_strings_array, /*is_closed_archive=*/ false);
+  guarantee(success, "shared strings array must point to only archivable objects");
```

Comment:
It should never fail. I could've used an assert but made it a guarantee since this is rather new code and I am paranoid.

Reply:
So what stops us from hitting this:
```
        // Don't archive a subgraph root that's too big. For archives static fields, that's OK
        // as the Java code will take care of initializing this field dynamically.
        return false;
```
in `HeapShared::archive_reachable_objects_from`?

Reply:
I have updated the code to make it structurally impossibly to reach this point, and added asserts to check for that. Explanations are here: 
```
void HeapShared::archive_strings() {
  oop shared_strings_array = StringTable::init_shared_table(_dumped_interned_strings);
  bool success = archive_reachable_objects_from(1, _default_subgraph_info, shared_strings_array, /*is_closed_archive=*/ false);
  // We must succeed because:
  // - _dumped_interned_strings do not contain any large strings.
  // - StringTable::init_shared_table() doesn't create any large arrays.
  assert(success, "shared strings array must not point to arrays or strings that are too large to archive");
  StringTable::set_shared_strings_array_index(append_root(shared_strings_array));
}
```

There are other asserts in the handling of `_dumped_interned_strings` and `shared_strings_array` to check their object sizes.